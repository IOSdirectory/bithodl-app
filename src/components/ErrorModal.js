import React, { Component } from 'react';
import { StyleSheet, View, TouchableHighlight, Text, LayoutAnimation, Dimensions } from 'react-native';
import PropTypes from 'prop-types';
import { VibrancyView } from 'react-native-blur';
import Icon from 'react-native-vector-icons/Ionicons';
import ReactNativeHaptic from 'react-native-haptic';

import StyledText from '../components/StyledText';
import Paragraph from '../components/Paragraph';

const windowDimensions = Dimensions.get('window');
const MODAL_MARGIN = 20;
const MODAL_WIDTH = windowDimensions.width - MODAL_MARGIN * 2;

const styles = StyleSheet.create({
  wrapper: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    zIndex: 2000
  },
  backdrop: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    opacity: 1
  },
  modal: {
    width: MODAL_WIDTH,
    backgroundColor: '#ffffff',
    borderRadius: 12,
    shadowColor: '#000000',
    shadowOpacity: 0.3,
    shadowOffset: { width: 0, height: 0 },
    shadowRadius: 12,
    marginTop: 0
  },
  modalHeader: {
    backgroundColor: '#FFD6D6',
    padding: 15,
    paddingTop: 17,
    borderTopLeftRadius: 12,
    borderTopRightRadius: 12
  },
  modalHeaderText: {
    color: '#FF5A59',
    fontSize: 16,
    fontWeight: '600',
    textAlign: 'center'
  },
  modalHeaderIcon: {
    fontSize: 28
  },
  modalBody: {
    padding: 25,
    paddingLeft: 30,
    paddingRight: 30
  },
  modalBodyText: {
    marginBottom: 0
  },
  modalButton: {
    padding: 15,
    alignItems: 'center',
    borderTopWidth: 1,
    borderColor: '#fbfbfb',
    borderBottomLeftRadius: 12,
    borderBottomRightRadius: 12
  },
  modalButtonText: {
    color: '#0073FF',
    fontWeight: '600',
    letterSpacing: 1
  }
});

export default class ErrorModal extends Component {
  componentWillReceiveProps(nextProps) {
    if (this.props.error !== nextProps.error) {
      LayoutAnimation.spring();

      if (nextProps.error) {
        ReactNativeHaptic.generate('impact');
      }
    }
  }

  render() {
    const error = this.props.error;
    const message = error ? error.message : '';
    const backdropStyles = [styles.backdrop];
    const modalStyles = [styles.modal];
    let pointerEvents = 'auto';

    if (!error) {
      pointerEvents = 'none';

      backdropStyles.push({
        opacity: 0
      });

      modalStyles.push({
        opacity: 0,
        marginTop: -500
      });
    }

    return (
      <View style={styles.wrapper} pointerEvents={pointerEvents}>
        <VibrancyView blurType='dark' blurAmount={5} style={backdropStyles} pointerEvents={pointerEvents} />

        <View style={modalStyles}>
          <View style={styles.modalHeader}>
            <StyledText style={styles.modalHeaderText}>
              <Icon name='ios-alert-outline' style={styles.modalHeaderIcon} />
            </StyledText>
          </View>

          <View style={styles.modalBody}>
            <Paragraph style={styles.modalBodyText}>{message}</Paragraph>
          </View>

          <View>
            <TouchableHighlight style={styles.modalButton} underlayColor='#fbfbfb' onPress={this.props.onDismiss}>
              <Text>
                <StyledText style={styles.modalButtonText}>OK</StyledText>
              </Text>
            </TouchableHighlight>
          </View>
        </View>
      </View>
    );
  }
}

ErrorModal.propTypes = {
  error: PropTypes.instanceOf(Error),
  onDismiss: PropTypes.func
};
