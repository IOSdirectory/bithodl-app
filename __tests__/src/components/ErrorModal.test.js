import React from 'react';
import renderer from 'react-test-renderer';
import ErrorModal from '../../../src/components/ErrorModal';

describe('ErrorModal', () => {
  it('renders correctly without an error', () => {
    const tree = renderer.create(
      <ErrorModal />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with an error', () => {
    const error = new Error('b4e5375f-a954-47bf-a0b5-6694394a042a');

    const tree = renderer.create(
      <ErrorModal error={error} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
