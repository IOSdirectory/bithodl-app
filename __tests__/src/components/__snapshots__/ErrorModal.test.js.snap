// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorModal renders correctly with an error 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": "center",
      "bottom": 0,
      "flex": 1,
      "justifyContent": "center",
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "zIndex": 2000,
    }
  }
>
  <VibrancyView
    blurAmount={5}
    blurType="dark"
    pointerEvents="auto"
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "opacity": 1,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
        ],
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ffffff",
          "borderRadius": 12,
          "marginTop": 0,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 12,
          "width": 710,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#FFD6D6",
          "borderTopLeftRadius": 12,
          "borderTopRightRadius": 12,
          "padding": 15,
          "paddingTop": 17,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        numberOfLines={undefined}
        style={
          Array [
            Object {
              "color": "#504F63",
              "fontFamily": "Avenir Next",
              "fontWeight": "400",
            },
            Object {
              "color": "#FF5A59",
              "fontSize": 16,
              "fontWeight": "600",
              "textAlign": "center",
            },
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 12,
              },
              Object {
                "fontSize": 28,
              },
              Object {
                "fontFamily": "Ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
      </Text>
    </View>
    <View
      style={
        Object {
          "padding": 25,
          "paddingLeft": 30,
          "paddingRight": 30,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        numberOfLines={undefined}
        style={
          Array [
            Object {
              "color": "#504F63",
              "fontFamily": "Avenir Next",
              "fontWeight": "400",
            },
            Array [
              Object {
                "color": "#8F8E94",
                "fontSize": 16,
                "marginBottom": 20,
                "textAlign": "center",
              },
              Object {
                "marginBottom": 0,
              },
            ],
          ]
        }
      >
        b4e5375f-a954-47bf-a0b5-6694394a042a
      </Text>
    </View>
    <View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            Object {
              "alignItems": "center",
              "borderBottomLeftRadius": 12,
              "borderBottomRightRadius": 12,
              "borderColor": "#fbfbfb",
              "borderTopWidth": 1,
              "padding": 15,
            },
          ]
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            numberOfLines={undefined}
            style={
              Array [
                Object {
                  "color": "#504F63",
                  "fontFamily": "Avenir Next",
                  "fontWeight": "400",
                },
                Object {
                  "color": "#0073FF",
                  "fontWeight": "600",
                  "letterSpacing": 1,
                },
              ]
            }
          >
            OK
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`ErrorModal renders correctly without an error 1`] = `
<View
  pointerEvents="none"
  style={
    Object {
      "alignItems": "center",
      "bottom": 0,
      "flex": 1,
      "justifyContent": "center",
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "zIndex": 2000,
    }
  }
>
  <VibrancyView
    blurAmount={5}
    blurType="dark"
    pointerEvents="none"
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "opacity": 1,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "opacity": 0,
          },
        ],
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ffffff",
          "borderRadius": 12,
          "marginTop": 0,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0.3,
          "shadowRadius": 12,
          "width": 710,
        },
        Object {
          "marginTop": -500,
          "opacity": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#FFD6D6",
          "borderTopLeftRadius": 12,
          "borderTopRightRadius": 12,
          "padding": 15,
          "paddingTop": 17,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        numberOfLines={undefined}
        style={
          Array [
            Object {
              "color": "#504F63",
              "fontFamily": "Avenir Next",
              "fontWeight": "400",
            },
            Object {
              "color": "#FF5A59",
              "fontSize": 16,
              "fontWeight": "600",
              "textAlign": "center",
            },
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 12,
              },
              Object {
                "fontSize": 28,
              },
              Object {
                "fontFamily": "Ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
        >
          
        </Text>
      </Text>
    </View>
    <View
      style={
        Object {
          "padding": 25,
          "paddingLeft": 30,
          "paddingRight": 30,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        numberOfLines={undefined}
        style={
          Array [
            Object {
              "color": "#504F63",
              "fontFamily": "Avenir Next",
              "fontWeight": "400",
            },
            Array [
              Object {
                "color": "#8F8E94",
                "fontSize": 16,
                "marginBottom": 20,
                "textAlign": "center",
              },
              Object {
                "marginBottom": 0,
              },
            ],
          ]
        }
      >
        
      </Text>
    </View>
    <View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            Object {
              "alignItems": "center",
              "borderBottomLeftRadius": 12,
              "borderBottomRightRadius": 12,
              "borderColor": "#fbfbfb",
              "borderTopWidth": 1,
              "padding": 15,
            },
          ]
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            numberOfLines={undefined}
            style={
              Array [
                Object {
                  "color": "#504F63",
                  "fontFamily": "Avenir Next",
                  "fontWeight": "400",
                },
                Object {
                  "color": "#0073FF",
                  "fontWeight": "600",
                  "letterSpacing": 1,
                },
              ]
            }
          >
            OK
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;
